packages/components/src/components/SquiggleViewer/ViewerProvider.tsx

```diff
@@ -219,26 +264,42 @@ export const ViewerProvider: FC<
   );
 
   const setCollapsed = (path: SqValuePath, isCollapsed: boolean) => {
-    const ref = settingsStoreRef.current[pathAsString(path)];
-    settingsStoreRef.current[pathAsString(path)] = {
-      ...ref,
-      collapsed: ref?.collapsed ?? isCollapsed,
-    };
+    setSettings(path, (state) => ({
+      ...state,
+      collapsed: state?.collapsed ?? isCollapsed,
+    }));
+  };
+
+  const updateCalculator = (
+    path: SqValuePath,
+    reduce: (calculator: CalculatorState) => CalculatorState
+  ) => {
+    const calculator = getSettingsRef(path).calculator;
+    if (calculator !== null) {
+      setSettings(path, (state) => ({
+        ...state,
+        calculator: reduce(calculator),
+      }));
+    }
   };
 
   const dispatch = useCallback(
     (action: Action) => {
```

Comment:
I'd probably extract this function and other local functions from this component.

- `getSettings`, `setSettings` and others can take `state` as an additional parameter;
- `dispatch` code could be moved to `const reducer: Reducer<SettingsStore, Action> = ...`;
- then the `ViewerProvider` component would only include one `useRef`, and `const dispatch = useCallback(action => reducer(settingsStoreRef.current, action), []);`.

For `getSettings` and `getMergedSettings` that are stored in the context you'd still have to create proxy functions with `useCallback`.

But this can be encapsulated in a custom hook: `const { getSettings, getMergedSettings, dispatch } = useSettingsStore();`

In general, hooks are great for managing the complexity like this.

(If you're not sure how to write that custom hook, I can help you with it. It's pretty straightforward, but I know can be unclear if you haven't done it before.)

Reply:
I think I'll leave this for later, seems like a more significant refactor.